File: java-common/src/net/devrieze/util/db/AbstractElementFactory.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with Foobar.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package net.devrieze.util.db;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;

import net.devrieze.util.db.DBHandleMap.HMElementFactory;


/**
 * Abstract baseclass for {@link ElementFactory} and {@link HMElementFactory}.
 * This class provides implementations that allow optional extensions for
 * complex datatypes. Simple types do not need factories that override these.
 *
 * @author Paul de Vrieze
 * @param <T> Type of the element created / handled
 */
public abstract class AbstractElementFactory<T> implements HMElementFactory<T> {

  @Override
  public CharSequence getFilterExpression() {
    return null;
  }

  @Override
  public int setFilterParams(PreparedStatement pStatement, int pOffset) throws SQLException {
    return 0;
  }

  @Override
  public void initResultSet(ResultSetMetaData pMetaData) throws SQLException {
    // Do nothing.
  }

  @Override
  public void postCreate(DBTransaction pConnection, T pElement) throws SQLException {
    // Do nothing.
  }

  @Override
  public void postStore(DBTransaction pConnection, long pHandle, T pOldValue, T pElement) throws SQLException {
    // Simple case, do nothing
  }

  @Override
  public void preRemove(DBTransaction pConnection, long pHandle) throws SQLException {
    // Don't do anything
  }

  @Override
  public void preRemove(DBTransaction pConnection, T pElement) throws SQLException {
    // Don't do anything
  }

  /**
   * Default implementation that just creates the element.
   */
  @Override
  public void preRemove(DBTransaction pConnection, ResultSet pElementSource) throws SQLException {
    preRemove(pConnection, create(pConnection, pElementSource));
  }

  @Override
  public void preClear(DBTransaction pConnection) throws SQLException {
    // Don't do anything
  }

}

File: java-common/src/main/kotlin/net/devrieze/util/db/AbstractElementFactory.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package net.devrieze.util.db

import net.devrieze.util.Handle
import uk.ac.bournemouth.kotlinsql.Column
import uk.ac.bournemouth.kotlinsql.Database
import uk.ac.bournemouth.kotlinsql.IColumnType
import uk.ac.bournemouth.util.kotlin.sql.DBConnection
import java.sql.SQLException


/**
 * Abstract baseclass for [ElementFactory] and [HMElementFactory].
 * This class provides implementations that allow optional extensions for
 * complex datatypes. Simple types do not need factories that override these.

 * @author Paul de Vrieze
 * *
 * @param  T of the element created / handled
 */
abstract class AbstractElementFactory<BUILDER, T:Any, TR:DBTransaction> : HMElementFactory<BUILDER, T, TR> {

  companion object {

    fun <T, S: IColumnType<T, S, C>, C:Column<T,S,C>> C.nullableValue(columns:List<Column<*,*,*>>, values:List<Any?>):T? {
      return values[columns.checkedIndexOf(this)]?.let{ type.cast(it) }
    }

    fun <T, S: IColumnType<T, S, C>, C:Column<T,S,C>> C.value(columns:List<Column<*,*,*>>, values:List<Any?>):T {
      return type.cast(values[columns.checkedIndexOf(this)]!!)
    }

    fun List<Column<*,*,*>>.checkedIndexOf(column:Column<*,*,*>): Int {
      return indexOf(column).also {
        if (it<0) throw SQLException("Column $column not found in $this")
      }
    }

  }

  override fun filter(select: Database._Where) = null

  @Throws(SQLException::class)
  override fun postStore(connection: DBConnection, handle: Handle<T>, oldValue: T?, newValue: T) {
    // Simple case, do nothing
  }

  @Throws(SQLException::class)
  override fun preRemove(transaction: TR, handle: Handle<T>) {
    // Don't do anything
  }

  @Throws(SQLException::class)
  override fun preRemove(transaction: TR, element: T) {
    // Don't do anything
  }

  override fun preRemove(transaction: TR, columns: List<Column<*, *, *>>, values: List<Any?>) {
    // Don't do anything
  }

  @Throws(SQLException::class)
  override fun preClear(transaction: TR) {
    // Don't do anything
  }

}
